homeassistant/components/airvisual_pro/sensor.py

```diff
@@ -0,0 +1,150 @@
+        key=SENSOR_KIND_AQI,
+        name="Air quality index",
+        device_class=SensorDeviceClass.AQI,
+        native_unit_of_measurement="AQI",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_BATTERY_LEVEL,
+        name="Battery",
+        device_class=SensorDeviceClass.BATTERY,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        native_unit_of_measurement=PERCENTAGE,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_CO2,
+        name="C02",
+        device_class=SensorDeviceClass.CO2,
+        native_unit_of_measurement=CONCENTRATION_PARTS_PER_MILLION,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_HUMIDITY,
+        name="Humidity",
+        device_class=SensorDeviceClass.HUMIDITY,
+        native_unit_of_measurement=PERCENTAGE,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_PM_0_1,
+        name="PM 0.1",
+        device_class=SensorDeviceClass.PM1,
+        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_PM_1_0,
+        name="PM 1.0",
+        device_class=SensorDeviceClass.PM10,
+        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_PM_2_5,
+        name="PM 2.5",
+        device_class=SensorDeviceClass.PM25,
+        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_TEMPERATURE,
+        name="Temperature",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=TEMP_CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_KIND_VOC,
+        name="VOC",
+        device_class=SensorDeviceClass.VOLATILE_ORGANIC_COMPOUNDS,
+        native_unit_of_measurement=CONCENTRATION_PARTS_PER_MILLION,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up AirVisual sensors based on a config entry."""
+    coordinator: DataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities(
+        AirVisualProSensor(coordinator, description)
+        for description in SENSOR_DESCRIPTIONS
+    )
+
+
+class AirVisualProSensor(AirVisualProEntity, SensorEntity):
+    """Define an AirVisual Pro sensor."""
+
+    _attr_has_entity_name = True
+
+    @callback
+    def _async_update_from_latest_data(self) -> None:
+        """Update the entity from the latest data."""
+        if self.entity_description.key == SENSOR_KIND_AQI:
```

Comment:
I'd probably make this a bit easier to read by making `settings` and `measurements` properties that are mapped to `self.coordinator.data["measurements"]` etc